pkg/reconcilermanager/controllers/reposync_controller_test.go

```diff
@@ -1806,6 +1807,129 @@ func validateGeneratedResourcesDeleted(t *testing.T, fakeClient *syncerFake.Clie
+	reqNamespacedName1 := namespacedName(rs1.Name, rs1.Namespace)
+
+	fakeClient, testReconciler := setupNSReconciler(t, rs1,
+		secretObj(t, reposyncSSHKey, configsync.AuthSSH, v1beta1.GitSource, core.Namespace(rs1.Namespace)))
+
+	// Test creating Deployment resources.
+	ctx := context.Background()
+	if _, err := testReconciler.Reconcile(ctx, reqNamespacedName1); err != nil {
+		t.Fatalf("unexpected reconciliation error, got error: %q, want error: nil", err)
+	}
+
+	label1 := map[string]string{
+		metadata.SyncNamespaceLabel: rs1.Namespace,
+		metadata.SyncNameLabel:      rs1.Name,
+		metadata.SyncKindLabel:      testReconciler.syncKind,
+	}
+
+	serviceAccount1 := fake.ServiceAccountObject(
+		nsReconcilerName,
+		core.Namespace(v1.NSConfigManagementSystem),
+		core.Labels(label1),
+	)
+	wantServiceAccounts := map[core.ID]*corev1.ServiceAccount{core.IDOf(serviceAccount1): serviceAccount1}
+
+	roleBinding1 := rolebinding(
+		RepoSyncPermissionsName(),
+		nsReconcilerName,
+		core.Namespace(rs1.Namespace),
+	)
+	wantRoleBindings := map[core.ID]*rbacv1.RoleBinding{core.IDOf(roleBinding1): roleBinding1}
+
+	repoContainerEnv1 := testReconciler.populateContainerEnvs(ctx, rs1, nsReconcilerName)
+	repoDeployment1 := repoSyncDeployment(nsReconcilerName,
+		setServiceAccountName(nsReconcilerName),
+		secretMutator(nsReconcilerName+"-"+reposyncSSHKey),
+		containerEnvMutator(repoContainerEnv1),
+	)
+	wantDeployments := map[core.ID]*appsv1.Deployment{core.IDOf(repoDeployment1): repoDeployment1}
+
+	if err := validateServiceAccounts(wantServiceAccounts, fakeClient); err != nil {
+		t.Error(err)
+	}
+	if err := validateRoleBindings(wantRoleBindings, fakeClient); err != nil {
+		t.Error(err)
+	}
+	if err := validateDeployments(wantDeployments, fakeClient); err != nil {
+		t.Errorf("Deployment validation failed. err: %v", err)
+	}
+	if t.Failed() {
+		t.FailNow()
+	}
+	t.Log("ServiceAccount, RoleBinding and Deployment successfully created")
+
+	// Add Finalizer
+	controllerutil.AddFinalizer(rs1, "test-finalizer")
+	if err := fakeClient.Update(ctx, rs1); err != nil {
+		t.Fatalf("failed to update the repo sync, got error: %v, want error: nil", err)
+	}
+
+	// Simulate a delete by adding a deletion timestamp
+	// TODO: Replace Update with Delete
+	// once the fake.Client has been updated to simulate deletions with finalizers.
+	now := metav1.Now()
+	rs1.DeletionTimestamp = &now
+	if err := fakeClient.Update(ctx, rs1); err != nil {
+		t.Fatalf("failed to update the repo sync, got error: %v, want error: nil", err)
+	}
+	if _, err := testReconciler.Reconcile(ctx, reqNamespacedName1); err != nil {
+		t.Fatalf("unexpected reconciliation error upon update, got error: %q, want error: nil", err)
+	}
+
+	// Validate nothing was updated
+	// TODO: Validate ResourceVersion has not been changed
+	// once the fake.Client has been updated to handle ResourceVersion simulation.
+	if err := validateServiceAccounts(wantServiceAccounts, fakeClient); err != nil {
```

Comment:
Would anything have been updated even if the deletion timestamp wasn't set? I assume nothing would change here since neither the resources nor the RepoSync spec has changed since the previous reconcile. I'm wondering if we need to change something to verify the update is ignored. Or am I missing something?

Reply:
My tests for ResourceVersion are in the fake.Client enhancement PR. Today we don't have any tests or way to prove that it hasn't actually updated because the fake.Client doesn't behave the same way the api server does. So I can't negative test it yet.

Reply:
Perhaps we could delete/modify an upserted resource or modify the RepoSync spec? I'm also fine with waiting until your other enhancement PR

Reply:
Fair, I could add a explicit update to ignore. Right now I guess it could pass just because of the DeepEqual skipping Updates when nothing changes. Tho ironically, we don't have a test for that yet either, because the fake.Client doesn't update ResourceVersion yet, so there's no way to prove it didn't happen yet.

Reply:
Done.

Reply:
I modified the RSync, to ignore the update, rather than modifying an updated resource, but both should be ignored.

Reply:
LGTM.

Minor nit is that these assertions on line 1893-1905 seem a bit redundant and could be rolled in with the assertions on line 1919-1931 with a single status update

Reply:
> Minor nit is that these assertions on line 1893-1905 seem a bit redundant and could be rolled in with the assertions on line 1919-1931 with a single status update

I wanted to make sure both cases were covered. Yes, they could be merged, but I kind of like having failures from a minimal set of changes to make debugging easier.